draft-irtf-cfrg-vdaf.md

```diff
@@ -3452,107 +3447,170 @@ In the remainder, we let `[n]` denote the set `{1, ..., n}` for positive integer
-(a share of) each gadget output. The verifier (share) consists of (a share of)
-the validity circuit's output and (a share of) each gadget test. The tests
-consume the query randomness."}
-
-On input of `meas`, `proof`, `query_rand`, and `joint_rand`, the verifier message
-is generated as follows:
-
-1. For every `i` in `[H]` create an empty table `wire_i`.
-
-1. Partition `proof` into the sub-vectors `seed_1`, `coeff_1`, ..., `seed_H`,
-   `coeff_H` defined in {{flp-bbcggi19-construction-prove}}.
-
-1. Evaluate `Valid` on input of `meas` and `joint_rand`, recording the inputs
-   of each gadget in the corresponding table. This step is similar to the
-   prover's step (3.) except the verifier does not evaluate the gadgets.
-   Instead, it computes the output of the `k`th call to `G_i` by evaluating
-   `poly_gadget_i(alpha_i^k)`. Let `out` denote the output of the circuit
-   evaluation.
-
-1. Next, reduce `out` as follows. If `EVAL_OUTPUT_LEN > 1`, then consume the
-   first element of `query_rand` by letting `[r], query_rand = front(1,
-   query_rand)`. Then let `v = r*out[0] + r**2*out[1] + r**3*out[2] + ...`.
-   That is, interpret the outputs as coefficients of a polynomial `f(x)` and
-   evaluate polynomial `f(x)*x` at a random point `r`.
-
-1. Compute the wire polynomials just as in the prover's step (4.).
-
-1. Compute the tests for well-formedness of the gadget polynomials. That is, for
-   every `i` in `[H]`:
-
-    * Let `t = query_rand[i]`. Check if `t^(P_i) == 1`: if so, then raise
-      ERR_ABORT and halt. (This prevents the verifier from inadvertently leaking
-      a gadget output in the verifier message.)
-
-    * Let `y_i = poly_gadget_i(t)`.
-
-    * For each `j` in `[0,L_i)` let `x_i[j-1] = poly_wire_i[j-1](t)`.
+{: title="Components of the query generation algorithm."}
+
+The query generation algorithm invokes the validity circuit on the encoded
+measurement and joint randomness. It evaluates the gadget polynomials
+encoded by the proof (share) to produce (a share of) each gadget output. The
+verifier (share) consists of (a share of) the validity circuit's output and (a
+share of) each gadget test. The gadget tests consume the query randomness.
+
+The goal of each gadget test is to ensure the inputs used by the prover to
+generate the gadget polynomial matches the inputs we used to evaluate it. We do
+this by partially reconstructing the gadget polynomial and evaluate it a random
+point: when we evaluate the gadget polynomial at the same point, we expect to
+get the same result.
+
+To start a gadget test, we first construct the (shares of the) wire polynomials
+just as the prover did. First, we record `g.wires[j][k]` as the input (share)
+of the `j`-th wire of the `k`-th invocation of the gadget. Again, this is
+accomplished by a wrapper gadget, `QueryGadget`, listed in {{gadget-wrappers}}.
+This gadget also evaluates the gadget proof for each gadget invocation. Then we
```

Comment:
Actually instead of "gadget proof", I think this should say "gadget polynomial". Does this make more sense?
```suggestion
This gadget also evaluates the gadget polynomial for each gadget invocation. Then we
```

Reply:
SGTM, elaborated a bit more here
```suggestion
This gadget also evaluates the gadget polynomial for each gadget invocation in order to produce the gadget's output. Then we
```